### ISSUE TITLE
state::change needs to be able to replace with a new node

### ISSUE THREAD

**Issue by alilee** [emmanuelantony2000/valerie#12]
The router works like an a singleton of an enum that represents the route options.

When the `onpopstate` event is received, a component local state (eg. `Atomic*`) will be set to the new route.

This requires that the state objects need to be able to replace themselves with an entirely new component, with their own children and channels. 

```rust
enum Routes {
    Login,
    Home,
    Post(PostID),
}

impl Default for Routes {
    fn default() -> Self {
        Self::Login
    }
} 

impl From<Url> for Routes {
    fn from(self) -> Routes {
        // ...
    }
}

fn router() -> Node {
    use Routes::*;
    let router = StateMutex<Routes>::new(); 
    HistoryService::route(router); // set up async monitor of the history API, convert url to Routes and send changes to `router`
    div!(
        router.formatter(|r| {
            match r {
                Login => login(),
                Home => home(),
                Post(id) => post(id),
            }
        }
    ).into()
}
```

What this requires is that StateMutex can take a closure returning a node which can replace the existing node.

Is this already in place, but I don't understand how to do this? or does it require an enhancement?

Any thoughts?

**Comment by alilee** [comment]
@emmanuelantony2000 Hiya, did you see this one?

**Comment by emmanuelantony2000** [comment]
I just got the notification after the last comment. 

So StateMutex currently takes the value, and then updates the node.

For the functionality you want, I was thinking of implementing StateUI, which should in theory take a closure returning a node, and then updates it, it could also be used for SSR and other stuff.

**Comment by alilee** [comment]
@emmanuelantony2000 

What do you think about separating this out into a more general Formatter/Presenter instance which would be constructed from a V, Receiver and closure, and replace the node when the V is updated down the Receiver? This would be the true "View" in MVVM. It would be constructed by passing a closure to States or Stores which would pull out the right V and Receiver:

```rust
fn router() -> Node {
    use Routes::*;
    let router = StateMutex<Routes>::new(); 
    HistoryService::route(router); // set up async monitor of the history API, convert url to Routes and send changes to `router`
    let parent: Node = div!().into();
    let formatter: Formatter<Routes> = router.formatter(parent, |r| {
            match r {
                Login => login(),
                Home => home(),
                Post(id) => post(id),
            }
        }
    )
    parent.push_child(formatter);
    parent
}
```

Complexity: It would need to know its parent Node and be able to replace itself in the parent's children, which means the Vec of children might not be the right answer for Node. Child nodes would need a unique id and would need to be in a Map.

**Comment by emmanuelantony2000** [comment]
This is good. In `valerie::Node` we could have a Weak link to the parent to track it. I could start working on implementing it, it could provide the base for a lot of other things I guess. And about the `Vec`, I think we could store the index of the Node with respect to the parent in the node itself?

For using a HashMap, we would need something unique, and hashable. Do tell if there is a way out to prevent the `Vec`.


Side note: I did some tests some hours before. We can't pass a Node downstream in a channel cause it doesn't implement `Sync`, but a reference of it.

**Comment by alilee** [comment]
I don't think you would pass the Node, only the model objects (behind State and Store), and that would ensure View separation.

The index might work, but would it be broken if the children change/add/remove?

**Comment by emmanuelantony2000** [comment]
Children add and remove will be done using the methods of `valerie::Node`. So the methods would ensure no breaking during runtime.
